go-controller/pkg/ovn/egressgw.go

```diff
@@ -368,13 +369,30 @@ func (oc *Controller) addPerPodGRSNAT(pod *kapi.Pod, podIfAddrs []*net.IPNet) er
 			// If the type is snat and the logical IP is the same, but external IP is different,
 			// even with --may-exist, the add may error out. this is because, for snat,
 			// (type, router, logical ip) is considered a key for uniqueness
-			stdout, stderr, err := util.RunOVNNbctl("--if-exists", "lr-nat-del", gr, "snat", podIP+mask,
-				"--", "lr-nat-add",
-				gr, "snat", gwIP, podIP+mask)
+			// Combining lr-nat-del and lr-nat-add will fail OVSDB transaction check because
+			// the entry will already exist so need to do these as two separate transactions
+			// Therefore we need to iterate through existing SNAT rules on GR and decide if we need to delete/add
+			// TODO(trozet) use go-bindings
+			stdout, stderr, err := util.RunOVNNbctl("lr-nat-list", gr)
```

Comment:
This search is to optimize the case where the snat rule to gwIP is already there.

I wonder if it's worth the effort (and the penalty in code readability), since `addPerPodGRSNAT` is called only when the lport is added the first time or if something changes in the ns annotation.

Reply:
If it was purely cosmetic then I dont think it is worth the effort. However, I'm seeing in the e2e control plane tests that add logical port actually fails:

```
217", FieldPath:""}): type: 'Warning' reason: 'ErrorAddingLogicalPort' failed to create SNAT rule for pod on gateway router GR_ovn-worker, stdout: "", stderr: "ovn-nbctl: 172.18.0.3, 10.244.2.3: a NAT with this external_ip and logical_ip already exists\n", error: OVN command '/usr/bin/ovn-nbctl --timeout=15 --if-exists lr-nat-del GR_ovn-worker snat 10.244.2.3/32 -- lr-nat-add GR_ovn-worker snat 172.18.0.3 10.244.2.3/32' failed: exit status 1
```

If this fails in add logical port we could end up failing to update other things about the port since we bail out of the function.